ambry-router/src/test/java/com.github.ambry.router/GetOperationTest.java

```diff
@@ -0,0 +1,474 @@
+/**
+ * Copyright 2016 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.router;
+
+import com.codahale.metrics.MetricRegistry;
+import com.github.ambry.clustermap.MockClusterMap;
+import com.github.ambry.commons.BlobId;
+import com.github.ambry.commons.ByteBufferReadableStreamChannel;
+import com.github.ambry.commons.LoggingNotificationSystem;
+import com.github.ambry.commons.ResponseHandler;
+import com.github.ambry.commons.ServerErrorCode;
+import com.github.ambry.config.RouterConfig;
+import com.github.ambry.config.VerifiableProperties;
+import com.github.ambry.messageformat.BlobInfo;
+import com.github.ambry.messageformat.BlobProperties;
+import com.github.ambry.network.NetworkClient;
+import com.github.ambry.network.NetworkClientErrorCode;
+import com.github.ambry.network.RequestInfo;
+import com.github.ambry.network.ResponseInfo;
+import com.github.ambry.protocol.RequestOrResponse;
+import com.github.ambry.utils.MockTime;
+import com.github.ambry.utils.Utils;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.Random;
+import java.util.concurrent.atomic.AtomicReference;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+
+
+public class GetOperationTest {
+  private static final int MAX_PORTS_PLAIN_TEXT = 3;
+  private static final int MAX_PORTS_SSL = 3;
+  private static final int CHECKOUT_TIMEOUT_MS = 1000;
+
+  private int requestParallelism = 2;
```

Comment:
Making variables final because they don't seem to be modified is sort of missing the point, don't you think? Somehow we have built up this coding culture that I don't think is correct. We should make variables `final` only if it does not make sense for them to be immutable - if they were constants. Most of what we point out in each other's reviews do fall into the constant category, but not all and I think that is because of the way we go about it - by checking if the variables are getting modified. Maybe they are not today, but they could be tomorrow?

In this case, the very reason `requestParallelism` is a variable is because I imagined being able to modify it. Otherwise, the value could be hardcoded when populating the properties. We don't have any tests that do that in this PR, but these tests could be extended.